cpp/src/arrow/compute/kernels/copy_data_internal.h

```diff
@@ -0,0 +1,109 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/compute/kernels/codegen_internal.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+template <typename Type, typename Enable = void>
+struct CopyDataUtils {};
+
+template <>
+struct CopyDataUtils<BooleanType> {
+  static void CopyData(const DataType&, const Scalar& in, const int64_t in_offset,
```

Comment:
Do the Scalar overloads need in_offset?

Reply:
`ReplaceWithArrayMaskImpl` in `vector_replace.cc` makes calls to `CopyData` using `ArrayData` and `Scalar` for the `Data` template type. When trying to select for which `CopyData` overload to call, there are errors in compilation when trying to differentiate the `ArrayData` version of `CopyData` from the Scalar version.

Reply:
Ah, I see now. Thanks.